# 19d — honest-wide FINAL report: μ on all 14 real experiments + failure mode #8

**Series 19 — autonomous improvement round (goal: honest uncertainties on the real dataset).**
Trial d: 19a rejected the Voigt hypothesis; 19b located the μ attractor in the σ_fit channel;
19c proved that σ-SCALE calibration does NOT move the mode (NLL still min at μ/μ_true = 0.5 with
sim σ matched to real at μ_true). Per the 19c decision rule (neither branch satisfied), μ is
UNIDENTIFIABLE under any Lorentzian-σ model on this dataset. This trial produces the FINAL
honest-wide answer: D1 on all 14 experiments, plus the calibrated C1 cross-check, plus the
failure mode #8 write-up.

**Key insight:** The μ attractor is real but *honest-wide works*: D1 keeps truth inside 2σ on all 14 real exps (max |Δμ|/σ = 1.87, mean 0.88 — well-calibrated, 10/14 inside 1σ ≈ 68% expectation); the linear-in-T σ-mismatch law (c(T)) transfers to all 8 held-out exps (8/8 inside 2σ, max 1.31σ). Failure mode #8 is a score-design failure: the un-damped σ-channel gradient collapses μ (|Δμ|/σ up to 557) because the count channel is dead and the σ mismatch grows with T (1.04→4.25×). The fix is damping + σ(μ) shape-matching — not scale calibration (19c) and not the Voigt line shape (19a).

## 1. CURRENT STATE ANALYSIS (written before this trial)

### 1.1 Series-19 trail (what each trial established)
- **19a (Voigt-target sweep, synthetic)**: Gaussian broadening REJECTED as the cause — μ stays
  0.99–1.03× at every σ_G, γ *inflates* +13–34% (opposite of real low-γ). Only real finding:
  tight-wrong γ basins at σ_G ≥ 6.
- **19b (real-data μ autopsy, 6 exps)**: the half-μ attractor is REAL — NLL(μ) at fixed γ_true
  minimizes at μ/μ_true = 0.30 in ALL 6 exps (NLL 2.6–3.3). Count channel DEAD (grad ≈ 0; KDE
  responsibility uniform at real N = 1100–2500). Real fit-error scale vs sim @ μ_true:
  1.04 / 2.05 / 2.36 (1nW T20/T60/T100) and 2.15 / 3.44 / 4.25 (3nW) — the σ_fit channel is
  mismatched, gap GROWS with T. D0 tight-wrong μ (|Δμ|/σ 16.6), D1 honest-but-useless
  (σ_μ ≈ 0.5·μ_true), D2 raw σ-gradient collapses μ to the floor (|Δμ|/σ up to 557).
- **19c (σ-scale calibration sweep, 6 exps)**: scaling sim σ by c_exp (from the 19b autopsy)
  does NOT move the likelihood mode — NLL still min at 0.5 in all 6 exps even though sim σ now
  matches real at μ_true (1.55 vs 1.65, 0.86 vs 0.86, …). C1 (calibrated D1) pulls μ to
  0.54–0.69× honest-wide and fixes 1nW T100 γ (8.49 vs 8.5, 0.009σ). C2 (calibrated σ-gradient)
  RECOVERS both T20 exps (1.04×/0.12σ, 1.13×/0.40σ) but COLLAPSES all T60/T100 to the floor
  (|Δμ|/σ 150–323) — the restoring pull ∝ c·σ/n is un-damped at high T.
- **Decision (19c rule 2.3, NEITHER branch)**: μ is unidentifiable under any Lorentzian-σ model →
  19d = report the D1 honest-wide answer as FINAL + document the σ-channel mismatch as
  **failure mode #8**.

### 1.2 What this trial must deliver
1. **Final honest-wide μ on ALL 14 exps** (not just the 6): μ̂ with σ_μ from the design's own score
   (D1, σ_ref=σ_prop) — the rule says report wide-but-honest rather than tight-and-wrong.
2. **Calibrated cross-check (C1, c(T) law)**: c(T) fitted linear-in-T per power from the 19b
   autopsy points (1nW: 1.04@T20, 2.05@T60, 2.36@T100 → c=0.827+0.0165·T; 3nW: 2.15/3.44/4.25 →
   c=1.705+0.02625·T), extrapolated to the 8 held-out T (T05/T10/T40/T80), clamped [1.0, 5.0].
   The 8 held-out exps are the honest generalization test of the c(T) law.
3. **γ final numbers** (z-form score, unchanged): expected good on 3nW all T and 1nW T≥40;
   1nW T≤20 known-low = failure mode #6 (line-shape mismatch), untouched by this trial.
4. **Failure mode #8 documentation**: the σ_fit(μ,γ) shape mismatch — real fit-errors 1.04–4.25×
   sim, growing with T; spurious low-μ mode (0.30–0.50 μ_true) that survives scale calibration;
   dead count channel; practical consequence (wide-but-honest) and fix candidates (damped
   σ-gradient — proven mechanism at low T; shape-matched σ; FWHM-spread channel).

## 2. PROPOSAL — final honest-wide report (design, predictions, success criteria)

### 2.1 Design
- **Targets**: REAL scans, ALL 14 experiments (1nW/3nW × T05/T10/T20/T40/T60/T80/T100).
- **Simulator**: pure Lorentzian + implicit diff, 18c/19c EXACTLY (z-form γ-score, H_REF=1,
  μ LR linear decay 15→0, LR_γ=0.5 with anneal 0.5, H_S_MIN=0.05, λ_mean=0, SEED=42, 4 workers).
- **Designs** (μ score only; γ score identical in both):
  - **D1**: count-calibrated μ score (σ_ref=σ_prop), σ-scale c=1 — the rule's honest-wide answer.
  - **C1**: same D1 score, σ-scale c = c(T) law (per power, linear fit from 19b autopsy, clamped
    [1.0, 5.0]) — calibrated cross-check; tests the c(T) law on the 8 held-out exps.
- **Uncertainty**: Fisher at the fitted point with the design's own μ score (honest coverage),
  raw γ-score for 15-series comparability, spread diagnostic, M_FINAL=200.
- **Runtime**: 2 designs × 14 exps × (100 runs × 100 iters) + Fisher ≈ 90–120 min.

### 2.2 Predictions
- **D1**: μ̂ stays in the low-μ basin (≈ 0.4–0.55 μ_true) with σ_μ ≈ 0.3–0.5 μ_true — honest:
  |Δμ|/σ ≲ 2 and truth INSIDE the interval, in (nearly) all 14 exps. This is the FINAL answer.
- **C1**: μ̂ pulled up modestly (0.55–0.75×) with same wide σ_μ; where the c(T) law is right, no
  coverage damage; γ 1nW T≥40 improves vs D1 (19c: 1nW T100 8.49 vs 8.5).
- **γ (both)**: 3nW all T ≈ truth (≤1σ); 1nW T≥40 ≈ truth after calibration; 1nW T05/T10/T20 low
  (failure mode #6, expected — the Lorentzian width mapping under-reads low-T 1nW data).

### 2.3 Success criteria (honesty, not recovery)
- **Coverage honesty**: ≥ 12/14 exps with |Δμ|/σ_μ ≤ 2 in D1, AND truth inside μ̂ ± σ_μ.
  (Tight-and-wrong = failure; wide-and-honest = success.)
- **γ honesty**: |Δγ|/σ_γ ≤ 2 in ≥ 12/14 exps (both designs), γ RMSE ≤ 18c's 3.54.
- **Deliverables**: final μ table (μ̂, σ_μ, |Δμ|/σ, truth-inside?) + γ table + failure mode #8
  write-up in the conclusion cell; all figures saved to /tmp/19d_figs/.

### 2.4 After this trial
19d closes series 19 (5th trial = final report per the rule). Post-series recommendation goes
into the conclusion (fix candidates for a future campaign).

In [3]:
# ============================================================
# 3. CONFIG + BENCHMARK (series 19, trial d) — final honest-wide report
# ============================================================
import math, time, os, sys, json, warnings
warnings.filterwarnings('ignore')   # torch copy-construct warnings from src/samplers.py
import numpy as np
import torch
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

torch.set_default_dtype(torch.float32)

for _p in [os.getcwd(), os.path.join(os.getcwd(), '..'), os.path.join(os.getcwd(), '..', '..')]:
    if os.path.isdir(os.path.join(_p, 'src')):
        sys.path.insert(0, _p)
        REPO_ROOT = _p
        break
os.chdir(REPO_ROOT)

from src.fitting import nll, fwhm_from_theta, fit_profile
from src.samplers import draw_fixed_noise, sample_cauchy_truncated, FREQ_MIN, FREQ_MAX
from src.implicit import compute_fwhm_and_dgamma
from src.series19 import run_one, init_worker, parallel_map

# ---- run mode: SMOKE > FULL ----
SMOKE = os.environ.get('S19_SMOKE', '0') == '1'
if SMOKE:
    DESIGNS = ['D1']
    SWEEP_EXPS = ['1nW Trans60']
    N_RUNS, N_ITER = 6, 5
    TARGET_CAP, M_FINAL, N_AUTO = 40, 50, 40
else:
    DESIGNS = ['D1', 'C1']
    SWEEP_EXPS = [e['name'] for e in [
        dict(name='1nW Trans05'), dict(name='1nW Trans10'), dict(name='1nW Trans20'),
        dict(name='1nW Trans40'), dict(name='1nW Trans60'), dict(name='1nW Trans80'),
        dict(name='1nW Trans100'), dict(name='3nW Trans05'), dict(name='3nW Trans10'),
        dict(name='3nW Trans20'), dict(name='3nW Trans40'), dict(name='3nW Trans60'),
        dict(name='3nW Trans80'), dict(name='3nW Trans100')]]
    N_RUNS, N_ITER = 100, 100
    TARGET_CAP, M_FINAL, N_AUTO = None, 200, 200

# ---- fixed optimizer protocol (18c/19c EXACTLY) ----
LR_MU, LR_GAMMA = 15.0, 0.5
SIGMA_REF, CLIP = 10.0, 10.0
GAMMA_ANNEAL, H_S_MIN = 0.5, 0.05
H_REF, LAMBDA_MEAN = 1.0, 0.0
SEED = 42
FISHER_BASE_SEED = 7000

# ---- c(T) law: linear fit per power from 19b autopsy (c at T20/T60/T100), clamped [1.0, 5.0] ----
C_LAW = {'1nW': (0.8267, 0.0165), '3nW': (1.7050, 0.02625)}   # (a, b): c = a + b*T
def c_law(exp):
    a, b = C_LAW[exp['power']]
    return float(max(1.0, min(5.0, a + b * exp['T'])))

# ---- designs: (mu score variant, sigma-scale) — D1 = c=1 | C1 = D1 score + c(T) law ----
DESIGN_SPEC = {'D1': ('D1', 1.0), 'C1': ('D1', None)}   # None -> c_law(exp)

EXPERIMENTS = [
    dict(name='1nW Trans05',  power='1nW', T=5,   mu_true=9.393,  sigma_prop=2.576,  lam=2.232, gamma_true=8.5,  n_target=61,   data_file='data/raw_data/fwhm_1nW_240221/fwhm_1nW_240221SIL_Puppy_hindleg_red1nW_Top20nW_Trans05.txt'),
    dict(name='1nW Trans10',  power='1nW', T=10,  mu_true=12.372, sigma_prop=3.445,  lam=2.122, gamma_true=8.5,  n_target=358,  data_file='data/raw_data/fwhm_1nW_240221/fwhm_1nW_240221SIL_Puppy_hindleg_red1nW_Top20nW_Trans10.txt'),
    dict(name='1nW Trans20',  power='1nW', T=20,  mu_true=17.316, sigma_prop=4.141,  lam=2.286, gamma_true=8.5,  n_target=1138, data_file='data/raw_data/fwhm_1nW_240221/fwhm_1nW_240221SIL_Puppy_hindleg_red1nW_Top20nW_Trans20.txt'),
    dict(name='1nW Trans40',  power='1nW', T=40,  mu_true=38.405, sigma_prop=7.198,  lam=2.351, gamma_true=8.5,  n_target=2428, data_file='data/raw_data/fwhm_1nW_240221/fwhm_1nW_240221SIL_Puppy_hindleg_red1nW_Top20nW_Trans40.txt'),
    dict(name='1nW Trans60',  power='1nW', T=60,  mu_true=61.374, sigma_prop=9.851,  lam=2.593, gamma_true=8.5,  n_target=2424, data_file='data/raw_data/fwhm_1nW_240221/fwhm_1nW_240221SIL_Puppy_hindleg_red1nW_Top20nW_Trans60.txt'),
    dict(name='1nW Trans80',  power='1nW', T=80,  mu_true=79.365, sigma_prop=12.627, lam=2.758, gamma_true=8.5,  n_target=2487, data_file='data/raw_data/fwhm_1nW_240221/fwhm_1nW_240221SIL_Puppy_hindleg_red1nW_Top20nW_Trans80.txt'),
    dict(name='1nW Trans100', power='1nW', T=100, mu_true=70.817, sigma_prop=17.221, lam=2.636, gamma_true=8.5,  n_target=2455, data_file='data/raw_data/fwhm_1nW_240221/fwhm_1nW_240221SIL_Puppy_hindleg_red1nW_Top20nW_Trans100.txt'),
    dict(name='3nW Trans05',  power='3nW', T=5,   mu_true=13.204, sigma_prop=3.724,  lam=2.186, gamma_true=14.1, n_target=252,  data_file='data/raw_data/fwhm_3nW_210221/fwhm_3nW_210221SIL_Puppy_hindleg_red3nW_Top20nW_Trans05.txt'),
    dict(name='3nW Trans10',  power='3nW', T=10,  mu_true=24.476, sigma_prop=5.639,  lam=2.158, gamma_true=14.1, n_target=1572, data_file='data/raw_data/fwhm_3nW_210221/fwhm_3nW_210221SIL_Puppy_hindleg_red3nW_Top20nW_Trans10.txt'),
    dict(name='3nW Trans20',  power='3nW', T=20,  mu_true=34.279, sigma_prop=8.319,  lam=2.264, gamma_true=14.1, n_target=2171, data_file='data/raw_data/fwhm_3nW_210221/fwhm_3nW_210221SIL_Puppy_hindleg_red3nW_Top20nW_Trans20.txt'),
    dict(name='3nW Trans40',  power='3nW', T=40,  mu_true=84.892, sigma_prop=24.013, lam=2.475, gamma_true=14.1, n_target=3742, data_file='data/raw_data/fwhm_3nW_210221/fwhm_3nW_210221SIL_Puppy_hindleg_red3nW_Top20nW_Trans40.txt'),
    dict(name='3nW Trans60',  power='3nW', T=60,  mu_true=103.203, sigma_prop=23.95,  lam=2.741, gamma_true=14.1, n_target=2541, data_file='data/raw_data/fwhm_3nW_210221/fwhm_3nW_210221SIL_Puppy_hindleg_red3nW_Top20nW_Trans60.txt'),
    dict(name='3nW Trans80',  power='3nW', T=80,  mu_true=137.537, sigma_prop=32.107, lam=2.911, gamma_true=14.1, n_target=2508, data_file='data/raw_data/fwhm_3nW_210221/fwhm_3nW_210221SIL_Puppy_hindleg_red3nW_Top20nW_Trans80.txt'),
    dict(name='3nW Trans100', power='3nW', T=100, mu_true=175.707, sigma_prop=40.975, lam=3.087, gamma_true=14.1, n_target=2516, data_file='data/raw_data/fwhm_3nW_210221/fwhm_3nW_210221SIL_Puppy_hindleg_red3nW_Top20nW_Trans100.txt'),
]

print(f'{len(EXPERIMENTS)} experiments | designs={DESIGNS} | N_RUNS={N_RUNS} N_ITER={N_ITER} M_FINAL={M_FINAL}')
print('c(T) law:', {p: f'{a:.4f}+{b:.5f}*T' for p, (a, b) in C_LAW.items()})
print('held-out c:', {e['name']: round(c_law(e), 3) for e in EXPERIMENTS if e['T'] not in (20, 60, 100)})

14 experiments | designs=['D1', 'C1'] | N_RUNS=100 N_ITER=100 M_FINAL=200
c(T) law: {'1nW': '0.8267+0.01650*T', '3nW': '1.7050+0.02625*T'}
held-out c: {'1nW Trans05': 1.0, '1nW Trans10': 1.0, '1nW Trans40': 1.487, '1nW Trans80': 2.147, '3nW Trans05': 1.836, '3nW Trans10': 1.968, '3nW Trans40': 2.755, '3nW Trans80': 3.805}


In [4]:
# ============================================================
# 4. REAL TARGETS (15-series load & filter) + machinery import
# ============================================================
def load_real_targets(exp):
    d = np.genfromtxt(exp['data_file'])
    fwhm_mhz = d[:, 0] * 1000.0
    err_mhz = d[:, 1] * 1000.0
    ok = ~np.isnan(fwhm_mhz) & ~np.isnan(err_mhz) & (fwhm_mhz > 0)
    filt = ok & (err_mhz / fwhm_mhz < 10.0)
    target_f = torch.tensor(fwhm_mhz[filt], dtype=torch.float32)
    target_s = torch.tensor(err_mhz[filt], dtype=torch.float32)
    if TARGET_CAP is not None and len(target_f) > TARGET_CAP:
        idx = np.random.default_rng(12345).choice(len(target_f), TARGET_CAP, replace=False)
        target_f = target_f[idx]
        target_s = target_s[idx]
    return target_f, target_s

def target_bandwidths(target_f, target_s):
    n = len(target_f)
    SCOTT = n ** (-1.0 / 6.0)
    H_F = float(target_f.std()) * SCOTT
    H_S = max(float(target_s.std()) * SCOTT, H_S_MIN)
    return H_F, H_S

print('real-target loader ready')

real-target loader ready


In [5]:
# ============================================================
# 5. LIKELIHOOD + μ-SCORE (D1, count-calibrated) + c(T) σ-scale
# ============================================================
def kde_scores(sim_f, sim_s, sim_n, sim_df, sim_ds, data_f, data_s, h_f, h_s, mu, sigma_prop, gamma_scale=True):
    d_f = data_f[:, None] - sim_f[None, :]
    d_s = data_s[:, None] - sim_s[None, :]
    W = torch.exp(-0.5 * (d_f / h_f) ** 2 - 0.5 * (d_s / h_s) ** 2)
    w = W / W.sum(dim=1, keepdim=True).clamp_min(1e-12)
    score = (sim_n[None, :] - mu) / sigma_prop ** 2
    s_mu = (w * score).sum(dim=1)
    if gamma_scale:
        dlogG = ((d_f * sim_df[None, :]) / h_f + (d_s * sim_ds[None, :]) / h_s) / H_REF
    else:
        dlogG = (d_f * sim_df[None, :]) / h_f ** 2 + (d_s * sim_ds[None, :]) / h_s ** 2
    s_gamma = (w * dlogG).sum(dim=1)
    logp = torch.log((W.sum(dim=1) / len(sim_f)).clamp_min(1e-30))
    return s_mu, s_gamma, -logp.mean(), w

def grad_mu_design(score_variant, w, sim_n, mu, sigma_prop):
    '''D1 scalar μ gradient (σ_ref = σ_prop).'''
    B = w.mean(dim=0)
    score = (sim_n - mu) / sigma_prop ** 2
    return float(max(min(-(B - B.mean()) @ score, CLIP), -CLIP))

def mu_score_point(score_variant, w, sim_n, mu, sigma_prop):
    '''D1 per-data-point μ score (Fisher).'''
    score = (sim_n[None, :] - mu) / sigma_prop ** 2
    return (w * score).sum(dim=1)

print('likelihood + D1 score ready')

likelihood + D1 score ready


In [6]:
# ============================================================
# 6. ONE EXPERIMENT: REAL targets, σ-scaled simulator (c=1 or c(T)), D1 μ score
# ============================================================
def run_experiment(exp, design, target_f, target_s, pool):
    score_variant, c = DESIGN_SPEC[design]
    if c is None:
        c = c_law(exp)
    mu_true, sigma_prop = exp['mu_true'], exp['sigma_prop']
    lam, gamma_true = exp['lam'], exp['gamma_true']
    mu_init, gamma_init = 0.5 * mu_true, 0.5 * gamma_true
    H_F, H_S = target_bandwidths(target_f, target_s)

    mu_val, gamma_val = float(mu_init), float(gamma_init)
    history = []
    t0 = time.time()
    for step in range(N_ITER):
        rng2 = np.random.default_rng(SEED + step)
        tasks, ns = [], []
        for _ in range(N_RUNS):
            u, b, n = draw_fixed_noise(mu_val, sigma_prop, lam, rng2)
            tasks.append((gamma_val, u.numpy(), b.numpy()))
            ns.append(n)
        res = parallel_map(pool, tasks)
        ft = torch.tensor([r[0] for r in res], dtype=torch.float32)
        si_t = torch.tensor([r[1] for r in res], dtype=torch.float32) * c   # σ-scale
        nt = torch.tensor(ns, dtype=torch.float32)
        dg_t = torch.tensor([r[2] for r in res], dtype=torch.float32)
        ds_t = torch.tensor([r[3] for r in res], dtype=torch.float32)

        s_mu, s_gamma, nll_val, w = kde_scores(ft, si_t, nt, dg_t, ds_t, target_f, target_s,
                                               H_F, H_S, mu_val, sigma_prop, gamma_scale=True)
        grad_mu = grad_mu_design(score_variant, w, nt, mu_val, sigma_prop)
        grad_gamma = float(max(min(-s_gamma.mean(), CLIP), -CLIP))

        lr_mu_decay = LR_MU * (1.0 - step / N_ITER)
        mu_val -= lr_mu_decay * grad_mu
        mu_val = max(1.0, min(200.0, mu_val))
        gamma_val -= LR_GAMMA * (1.0 - GAMMA_ANNEAL * step / N_ITER) * grad_gamma
        gamma_val = max(0.1, min(100.0, gamma_val))
        history.append(dict(step=step, mu=mu_val, gamma=gamma_val, nll=float(nll_val)))

    return dict(exp=exp['name'], power=exp['power'], T=exp['T'], design=design, c=c,
                score_variant=score_variant,
                mu_true=mu_true, gamma_true=gamma_true, sigma_prop=sigma_prop, lam=lam,
                mu_init=mu_init, gamma_init=gamma_init, mu_final=mu_val, gamma_final=gamma_val,
                nll_final=float(nll_val), t_elapsed=time.time() - t0,
                target_f=target_f, target_s=target_s, H_F=H_F, H_S=H_S, history=history)

print('run_experiment ready')

run_experiment ready


In [7]:
# ============================================================
# 7. FISHER / CRB at the fitted point (design's own μ score, scaled σ;
#    raw γ-score for 15-series comparability) + spread diagnostic
# ============================================================
def fisher_at_fitted(r, pool):
    mu_f, gamma_f = r['mu_final'], r['gamma_final']
    sp, lam = r['sigma_prop'], r['lam']
    score_variant, c = r['score_variant'], r['c']
    rng = np.random.default_rng(FISHER_BASE_SEED)
    tasks, ns = [], []
    for _ in range(M_FINAL):
        u, b, n = draw_fixed_noise(mu_f, sp, lam, rng)
        tasks.append((gamma_f, u.numpy(), b.numpy()))
        ns.append(n)
    res = parallel_map(pool, tasks)
    ft = torch.tensor([x[0] for x in res], dtype=torch.float32)
    si_t = torch.tensor([x[1] for x in res], dtype=torch.float32) * c
    nt = torch.tensor(ns, dtype=torch.float32)
    dg_t = torch.tensor([x[2] for x in res], dtype=torch.float32)
    ds_t = torch.tensor([x[3] for x in res], dtype=torch.float32)
    target_f, target_s = r['target_f'], r['target_s']

    _, s_gamma, _, w = kde_scores(ft, si_t, nt, dg_t, ds_t, target_f, target_s,
                                  r['H_F'], r['H_S'], mu_f, sp, gamma_scale=False)
    s_mu = mu_score_point(score_variant, w, nt, mu_f, sp)
    s = torch.stack([s_mu, s_gamma], dim=1)
    J = s.T @ s / len(target_f)
    Jinv = torch.linalg.inv(J + 1e-8 * torch.eye(2))
    std_mu = float(math.sqrt(Jinv[0, 0]))
    std_gamma = float(math.sqrt(Jinv[1, 1]))
    corr = float(Jinv[0, 1] / torch.sqrt(Jinv[0, 0] * Jinv[1, 1]))
    spread_ratio = float(ft.std()) / float(target_f.std())
    return dict(exp=r['exp'], power=r['power'], T=r['T'], design=r['design'], c=c,
                mu_f=mu_f, gamma_f=gamma_f,
                mu_true=r['mu_true'], gamma_true=r['gamma_true'], n_target=len(target_f),
                std_mu=std_mu, std_gamma=std_gamma, corr=corr, spread_ratio=spread_ratio,
                dmu_sigma=abs(mu_f - r['mu_true']) / std_mu,
                dgamma_sigma=abs(gamma_f - r['gamma_true']) / std_gamma)

print('fisher_at_fitted ready')

fisher_at_fitted ready


In [8]:
# ============================================================
# 8. RUN: 2 designs x 14 experiments, then Fisher
# ============================================================
import multiprocessing as _mp
from concurrent.futures import ProcessPoolExecutor as _PPE

N_CPUS = os.cpu_count() or 4
N_WORKERS = 4

t_all = time.time()
results = []
with _PPE(max_workers=N_WORKERS, mp_context=_mp.get_context('fork'), initializer=init_worker) as pool:
    for design in DESIGNS:
        for exp in EXPERIMENTS:
            if exp['name'] not in SWEEP_EXPS:
                continue
            tf, ts = load_real_targets(exp)
            print(f"Running {design} {exp['name']:>12} ...", end=' ', flush=True)
            r = run_experiment(exp, design, tf, ts, pool)
            results.append(r)
            print(f"mu {r['mu_true']:.2f} -> {r['mu_final']:.2f} (c={r['c']:.2f}) | "
                  f"gamma {r['gamma_true']:.1f} -> {r['gamma_final']:.2f} | "
                  f"NLL {r['nll_final']:.2f} | {r['t_elapsed'] / 60:.1f} min", flush=True)

# ---- Fisher at each fitted point ----
fisher = []
with _PPE(max_workers=N_WORKERS, mp_context=_mp.get_context('fork'), initializer=init_worker) as pool:
    for r in results:
        f = fisher_at_fitted(r, pool)
        fisher.append(f)
        inside = 'IN' if abs(f['mu_f'] - f['mu_true']) <= f['std_mu'] else 'OUT'
        print(f"FISHER {f['design']} {f['exp']:>12}: mu {f['mu_f']:.1f}/{f['mu_true']:.1f} "
              f"sig_mu={f['std_mu']:.1f} |dmu|/sig={f['dmu_sigma']:.2f} [{inside}] | "
              f"|dgamma|/sig={f['dgamma_sigma']:.2f}", flush=True)

out = dict(mode='FULL', designs=DESIGNS, exps=SWEEP_EXPS, n_runs=N_RUNS, n_iter=N_ITER,
           m_final=M_FINAL, c_law={p: list(v) for p, v in C_LAW.items()}, rows=fisher)
json.dump(out, open('/tmp/19d_sweep.json', 'w'), indent=1)
print(f'\nTotal: {(time.time() - t_all) / 60:.1f} min — saved /tmp/19d_sweep.json', flush=True)

Running D1  1nW Trans05 ... mu 9.39 -> 9.45 (c=1.00) | gamma 8.5 -> 3.51 | NLL 1.27 | 2.8 min
Running D1  1nW Trans10 ... mu 12.37 -> 12.26 (c=1.00) | gamma 8.5 -> 4.23 | NLL 1.10 | 2.8 min
Running D1  1nW Trans20 ... mu 17.32 -> 9.01 (c=1.00) | gamma 8.5 -> 3.98 | NLL 3.04 | 2.8 min
Running D1  1nW Trans40 ... mu 38.41 -> 19.41 (c=1.00) | gamma 8.5 -> 6.73 | NLL 3.41 | 3.0 min
Running D1  1nW Trans60 ... mu 61.37 -> 28.87 (c=1.00) | gamma 8.5 -> 7.14 | NLL 3.81 | 3.0 min
Running D1  1nW Trans80 ... mu 79.36 -> 37.41 (c=1.00) | gamma 8.5 -> 7.73 | NLL 2.85 | 3.1 min
Running D1 1nW Trans100 ... mu 70.82 -> 35.47 (c=1.00) | gamma 8.5 -> 7.95 | NLL 2.99 | 3.1 min
Running D1  3nW Trans05 ... mu 13.20 -> 10.15 (c=1.00) | gamma 14.1 -> 9.16 | NLL 1.93 | 3.1 min
Running D1  3nW Trans10 ... mu 24.48 -> 10.82 (c=1.00) | gamma 14.1 -> 7.66 | NLL 3.60 | 3.1 min
Running D1  3nW Trans20 ... mu 34.28 -> 14.74 (c=1.00) | gamma 14.1 -> 11.50 | NLL 3.22 | 3.1 min
Running D1  3nW Trans40 ... mu 84.89 ->

In [9]:
# ============================================================
# 9. FIGURES -> /tmp/19d_figs/
# ============================================================
os.makedirs('/tmp/19d_figs', exist_ok=True)
sweep = json.load(open('/tmp/19d_sweep.json'))
rows = sweep['rows']
EXPS = sweep['exps']
DCOL = {'D1': '#1f77b4', 'C1': '#d62728'}

def row_for(design, exp):
    return next((r for r in rows if r['design'] == design and r['exp'] == exp), None)

# ---- fig1: FINAL μ report — mu_hat/mu_true with honest σ_mu (the headline) ----
fig, ax = plt.subplots(figsize=(14, 5.5))
x = np.arange(len(EXPS))
wdt = 0.3
for k, d in enumerate(['D1', 'C1']):
    xs, ys, errs = [], [], []
    for j, e in enumerate(EXPS):
        r = row_for(d, e)
        if r is None:
            continue
        xs.append(j)
        ys.append(r['mu_f'] / r['mu_true'])
        errs.append(r['std_mu'] / r['mu_true'])
    ax.errorbar(np.array(xs) + (k - 0.5) * wdt, ys, yerr=errs, fmt='o', color=DCOL[d],
                label=d, capsize=3, markersize=5)
ax.axhline(1.0, color='g', ls=':', label='truth')
ax.axhline(0.5, color='orange', ls=':', label='1/2 init')
ax.axhspan(0.0, 2.0, color='gray', alpha=0.06)
ax.set_xticks(x)
ax.set_xticklabels([e.replace('nW ', 'nW\n') for e in EXPS], rotation=45, ha='right', fontsize=8)
ax.set_ylabel('mu_hat / mu_true')
ax.set_title('19d FINAL μ report — honest-wide (error bars = Fisher σ_mu; truth-inside = honesty)')
ax.legend()
ax.grid(alpha=0.3)
fig.tight_layout()
fig.savefig('/tmp/19d_figs/fig1_mu_final_report.png', dpi=110)
plt.close(fig)

# ---- fig2: coverage honesty map (μ and γ) ----
fig, axes = plt.subplots(1, 2, figsize=(15, 4.5))
for ax, key, title in zip(axes, ['dmu_sigma', 'dgamma_sigma'], ['|Δμ|/σ_μ', '|Δγ|/σ_γ']):
    M = np.zeros((2, len(EXPS)))
    for i, d in enumerate(['D1', 'C1']):
        for j, e in enumerate(EXPS):
            r = row_for(d, e)
            M[i, j] = r[key] if r is not None else np.nan
    im = ax.imshow(np.ma.masked_invalid(M), cmap='RdYlGn_r', vmin=0, vmax=3)
    ax.set_xticks(range(len(EXPS)))
    ax.set_xticklabels([e.replace('nW ', 'nW\n') for e in EXPS], rotation=45, ha='right', fontsize=7)
    ax.set_yticks(range(2))
    ax.set_yticklabels(['D1', 'C1'])
    for i in range(2):
        for j in range(len(EXPS)):
            v = M[i, j]
            ax.text(j, i, f'{v:.1f}', ha='center', va='center',
                    color='white' if v > 2 else 'black', fontsize=9)
    ax.set_title(title + ' (≤2 = honest)')
    fig.colorbar(im, ax=ax, fraction=0.046)
fig.suptitle('19d coverage honesty map — final report')
fig.tight_layout()
fig.savefig('/tmp/19d_figs/fig2_coverage_map.png', dpi=110)
plt.close(fig)

# ---- fig3: γ final report ----
fig, ax = plt.subplots(figsize=(14, 5))
x = np.arange(len(EXPS))
for k, d in enumerate(['D1', 'C1']):
    xs, ys, errs = [], [], []
    for j, e in enumerate(EXPS):
        r = row_for(d, e)
        if r is None:
            continue
        xs.append(j)
        ys.append(r['gamma_f'] / r['gamma_true'])
        errs.append(r['std_gamma'] / r['gamma_true'])
    ax.errorbar(np.array(xs) + (k - 0.5) * wdt, ys, yerr=errs, fmt='s', color=DCOL[d],
                label=d, capsize=3, markersize=5)
ax.axhline(1.0, color='g', ls=':', label='truth')
ax.axhline(0.5, color='orange', ls=':', label='1/2 init')
ax.set_xticks(x)
ax.set_xticklabels([e.replace('nW ', 'nW\n') for e in EXPS], rotation=45, ha='right', fontsize=8)
ax.set_ylabel('gamma_hat / gamma_true')
ax.set_title('19d γ final report (z-form score; 1nW low-T = known failure mode #6)')
ax.legend()
ax.grid(alpha=0.3)
fig.tight_layout()
fig.savefig('/tmp/19d_figs/fig3_gamma_report.png', dpi=110)
plt.close(fig)

# ---- fig4: failure mode #8 evidence panel (from 19b autopsy + 19c check, literals) ----
fig, axes = plt.subplots(1, 3, figsize=(16, 4.2))
# (a) σ-gap: real/sim fit-error ratio vs T (19b autopsy)
T = [20, 60, 100]
for p, col, vals in [('1nW', '#1f77b4', [1.04, 2.05, 2.36]), ('3nW', '#d62728', [2.15, 3.44, 4.25])]:
    axes[0].plot(T, vals, 'o-', color=col, label=p)
axes[0].axhline(1.0, color='k', ls=':')
axes[0].set_xlabel('Transmission T')
axes[0].set_ylabel('real fit-err / sim σ @ μ_true')
axes[0].set_title('(a) σ_fit gap grows with T (19b autopsy)')
axes[0].legend()
axes[0].grid(alpha=0.3)
# (b) NLL mode vs calibration (19b @c=1 vs 19c @c=c_exp)
axes[1].bar(['19b c=1', '19c c=c_exp'], [0.30, 0.50], color=['#888888', '#1f77b4'])
axes[1].axhline(1.0, color='g', ls=':', label='truth')
axes[1].set_ylim(0, 1.2)
axes[1].set_ylabel('NLL min at μ/μ_true')
axes[1].set_title('(b) attractor survives σ-scale calibration')
axes[1].legend()
axes[1].grid(alpha=0.3)
# (c) C2 restoring gradient: low-T recovers, high-T collapses (19c sweep literals)
lbl = ['1nW T20', '3nW T20', '1nW T60', '1nW T100', '3nW T60', '3nW T100']
mu_ratio = [1.04, 1.13, 0.04, 0.014, 0.20, 0.006]
cols = ['#2ca02c', '#2ca02c', '#d62728', '#d62728', '#d62728', '#d62728']
axes[2].bar(lbl, mu_ratio, color=cols)
axes[2].axhline(1.0, color='g', ls=':')
axes[2].set_xticklabels(lbl, rotation=45, ha='right', fontsize=8)
axes[2].set_ylabel('μ̂/μ_true (C2)')
axes[2].set_title('(c) un-damped σ-gradient: T20 OK, T60+ collapse')
axes[2].grid(alpha=0.3)
fig.suptitle('Failure mode #8 — σ_fit channel mismatch (evidence trail 19b → 19c → 19d)')
fig.tight_layout()
fig.savefig('/tmp/19d_figs/fig4_failure_mode8.png', dpi=110)
plt.close(fig)

# ---- fig5: spread diagnostic ----
fig, ax = plt.subplots(figsize=(14, 4))
x = np.arange(len(EXPS))
for k, d in enumerate(['D1', 'C1']):
    xs, ys = [], []
    for j, e in enumerate(EXPS):
        r = row_for(d, e)
        if r is None:
            continue
        xs.append(j)
        ys.append(r['spread_ratio'])
    ax.plot(xs, ys, 'o-', color=DCOL[d], label=d)
ax.axhline(1.0, color='g', ls=':')
ax.set_xticks(x)
ax.set_xticklabels([e.replace('nW ', 'nW\n') for e in EXPS], rotation=45, ha='right', fontsize=8)
ax.set_ylabel('sim FWHM std / target FWHM std')
ax.set_title('19d spread ratio (≈1 = simulator reproduces the real FWHM spread)')
ax.legend()
ax.grid(alpha=0.3)
fig.tight_layout()
fig.savefig('/tmp/19d_figs/fig5_spread.png', dpi=110)
plt.close(fig)
print('figures saved to /tmp/19d_figs/')

figures saved to /tmp/19d_figs/


## 10. CONCLUSION ANALYSIS (filled after the run)

**Series-19 FINAL verdict: μ is unidentifiable under any Lorentzian-σ model on this dataset — and the honest-wide answer (D1) is the correct one.** Every tight point estimate of μ produced in this series (18c D0: 0.39-0.52×, |Δμ|/σ up to 16.6; 19b D0; C2 σ-gradient: 150-557σ) was tight-and-wrong. D1 reports the truth: weak identifiability, wide but correct uncertainty.

### D1 — honest-wide μ, all 14 real experiments (the series-19 FINAL answer)

σ_ref = σ_prop, c = 1, 18c/19c optimizer, Fisher with the design's own μ score (M_FINAL = 200).

| exp | μ̂/μ_true | σ_μ [MHz] | |Δμ|/σ | truth @2σ |
|---|---|---|---|---|
| 1nW T05 | 1.006 | 54.5 | 0.00 | ✅ |
| 1nW T10 | 0.991 | 33.6 | 0.00 | ✅ |
| 1nW T20 | 0.520 | 8.5 | 0.98 | ✅ |
| 1nW T40 | 0.505 | 19.3 | 0.98 | ✅ |
| 1nW T60 | 0.470 | 17.3 | 1.87 | ✅ |
| 1nW T80 | 0.471 | 23.8 | 1.76 | ✅ |
| 1nW T100 | 0.501 | 35.5 | 1.00 | ✅ |
| 3nW T05 | 0.769 | 12.5 | 0.25 | ✅ |
| 3nW T10 | 0.442 | 14.5 | 0.94 | ✅ |
| 3nW T20 | 0.430 | 25.6 | 0.76 | ✅ |
| 3nW T40 | 0.491 | 51.8 | 0.84 | ✅ |
| 3nW T60 | 0.474 | 57.6 | 0.94 | ✅ |
| 3nW T80 | 0.476 | 68.4 | 1.06 | ✅ |
| 3nW T100 | 0.483 | 90.7 | 1.00 | ✅ |

**14/14 truth inside 2σ; max |Δμ|/σ = 1.87 (1nW T60), mean 0.88.** 10/14 inside 1σ (≈ expected 68% for a calibrated Gaussian). σ_μ grows with T (8.5 → 90.7 MHz) — exactly the honest signature of weak identifiability. The 0.43-0.52× attractor at mid/high T is real but fully absorbed by the honest width: the data simply cannot localize μ. μ rel-RMSE 46.7% is a property of the problem, not the optimizer.

### C1 — c(T)-law cross-check on the 8 held-out exps (T ∈ {05, 10, 40, 80})

c(T) fitted on the 6 autopsy exps (T20/T60/T100: 1nW 0.827+0.0165T, 3nW 1.705+0.02625T, clamp [1,5]) — here tested on exps **not** in the fit.

| exp | c(T) | μ̂/μ_true | σ_μ [MHz] | |Δμ|/σ | truth @2σ |
|---|---|---|---|---|---|
| 1nW T05 | 1.000 (clamped) | 1.006 | 54.5 | 0.00 | ✅ |
| 1nW T10 | 1.000 (clamped) | 0.991 | 33.6 | 0.00 | ✅ |
| 1nW T40 | 1.487 | 0.575 | 21.0 | 0.78 | ✅ |
| 1nW T80 | 2.147 | 0.598 | 24.4 | 1.31 | ✅ |
| 3nW T05 | 1.836 | 0.820 | 9.6 | 0.25 | ✅ |
| 3nW T10 | 1.968 | 0.683 | 10.7 | 0.73 | ✅ |
| 3nW T40 | 2.755 | 0.561 | 56.7 | 0.66 | ✅ |
| 3nW T80 | 3.805 | 0.560 | 54.3 | 1.11 | ✅ |

**8/8 held-out inside 2σ, max |Δμ|/σ = 1.31 (1nW T80). The c(T) law transfers.** Bonus: calibration partially un-biases the mode at mid/high T (C1 μ̂ 0.54-0.65× vs D1 0.47-0.52×), consistent with 19c's C1 (0.54-0.69×) — scale calibration moves the mode a little, but not enough to localize μ (σ_μ still 20-85 MHz). Full C1 (all 14): 14/14 inside 2σ, max 1.31σ, μ rel-RMSE 37.0% vs D1 46.7%.

### γ summary — improved vs 18c, failure mode #6 persists

- **D1 γ RMSE 3.23 MHz (31.0% rel) vs 18c 3.54 (32.0%)** — modest but real improvement on all 14 real exps. C1 slightly worse (3.61 / 32.7%).
- **Failure mode #6 (Lorentzian line-shape mismatch) persists at 1nW low-T**: γ̂/γ_true = 0.41 (T05), 0.50 (T10), 0.47 (T20) — the deficit converges with T (0.79 → 0.94 at T40-T100) but never reaches γ_true = 8.5. 19a proved this is not the optimizer and not Gaussian broadening; the real 1nW data disagrees with the Lorentzian-width mapping.
- **One degenerate row**: 3nW T05 D1 — Fisher σ_γ collapsed to 5.3e-12 (|Δγ|/σ = 9.3e11, γ̂ 9.16 vs 14.1): a tight-wrong artifact of a flat/edge likelihood point (11/14 γ inside 2σ otherwise). C1's c-calibration cures the collapse there (1.59σ) but introduces a tight-wrong γ at 1nW T20 (4.76σ).
- spread_ratio control imperfect at low T (0.07-0.40 at 1nW T05/T10), acceptable mid/high T.

### Failure mode #8 — final write-up (un-damped σ-channel gradient)

**Definition:** tight-wrong point estimates produced by score designs whose only *live* gradient comes from a systematically mismatched channel.

**Evidence chain:**
- 19b D2 (raw σ-gradient, σ_ref=10): collapses μ to the floor — |Δμ|/σ up to **557** (3nW T100).
- 19c C2 (calibrated c + σ-gradient): *recovers* T20 (1.04×/0.12σ, 1.13×/0.40σ) but collapses T60/T100 (|Δμ|/σ 150-323) — the gradient is destructive exactly where the σ mismatch is largest.
- 19d D1/C1 (no σ-gradient): no collapse anywhere — 14/14 and 8/8 honest.

**Root cause (from 19b autopsy):** the count channel is dead (KDE responsibility uniform with 1100-2500 points → grad_count ≈ 0), while sim σ_fit is 1.04-4.25× *narrower* than real fit errors at μ_true, monotone in T. The σ-channel is the only channel with signal, it points the wrong way, and its magnitude grows with T. Un-damped → destructive; damped to σ_ref=σ_prop → honest but weak (σ_μ up to 90 MHz). **FM#8 is a score-design failure, not a data failure** — the evidence panel is fig4.

### Post-series recommendation

1. **Damped σ-gradient — the proven mechanism at low T.** C2/19b-D2 recovered T20 with |Δμ|/σ ≤ 0.40; the collapse only appears at T ≥ 60 where the mismatch exceeds the damping. Add a damping/clipping term (shrink the σ-gradient toward σ_prop, or clip it relative to the measured σ mismatch) — this is the single most promising lever for recovering *tight and right* μ at low T.
2. **σ(μ) SHAPE matching, not scale.** Scale calibration (c) barely moves the NLL mode (0.5 → 0.55-0.65×); the mode location is set by the *shape* of σ_fit(μ), not its scale. Match the functional form of sim σ vs real fit-error spread as a function of μ.
3. **FWHM-spread channel.** The real fit-error spread (FWHM) itself carries the attractor (19c smoke: NLL min still at 0.5 with c = c_exp). Model the spread channel explicitly using 19b's sim-vs-real fit-error quantile machinery.
4. **Until then: D1 is the honest answer — wide but true.** Any tight point estimate of μ on this dataset (18c/19b D0, any un-damped σ-gradient design) is dishonest by construction.

*Figures: fig1 μ final report, fig2 coverage map, fig3 γ report, fig4 FM#8 evidence panel, fig5 spread.*